File: src/doody/doody_bringup/scripts/t265_odom.py
#!/usr/bin/env python3
 
import math
import random
from copy import deepcopy

import rclpy
from rclpy.node import Node
from rclpy.clock import Clock, ClockType
from geometry_msgs.msg import Pose, Twist
from nav_msgs.msg import Odometry

#!/usr/bin/env python3
# /home/ws/src/doody/doody_bringup/scripts/t265_odom.py
# ROS2 Humble node: subscribe to geometry_msgs/Pose and publish nav_msgs/Odometry



class PoseToOdom(Node):
    def __init__(self):
        super().__init__('t265_odom_publisher')

        # configurable parameters
        self.declare_parameter('input_topic', 'sensors/gt') #directly from gazebo is delayed and scale ?!? to implement in Jazzy
        #self.declare_parameter('input_topic', 'platform/odom')
        self.declare_parameter('output_topic', 'sensors/t265/pose/sample')
        self.declare_parameter('odom_frame', 'odom')
        self.declare_parameter('base_frame', 't265_link')
        # standard deviation for per-field gaussian noise applied to pose
        # small default so behavior changes only slightly; adjust via ros2 param
        self.declare_parameter('position_noise_std', 0.005)
        self.declare_parameter('orientation_noise_std', 0.0001)

        input_topic = self.get_parameter('input_topic').get_parameter_value().string_value
        output_topic = self.get_parameter('output_topic').get_parameter_value().string_value
        self.odom_frame = self.get_parameter('odom_frame').get_parameter_value().string_value
        self.base_frame = self.get_parameter('base_frame').get_parameter_value().string_value
        self.odom_pub = self.create_publisher(Odometry, output_topic, 1)
        self.pose_sub = self.create_subscription(Pose, input_topic, self.pose_callback, 1)
        
        self.initialised = False
        self.offset = Pose()

        # storage for latest received pose and computed velocity
        self.latest_pose = None
        self._prev_pose = None
        self._prev_pose_time_ns = None
        self.latest_vel = None

        # publish rate (Hz) and timer to publish odom regularly
        self.declare_parameter('publish_rate', 20.0)
        self.publish_rate = float(self.get_parameter('publish_rate').get_parameter_value().double_value)
        self.create_timer(1.0 / max(1e-6, self.publish_rate), self.run)

        self.get_logger().info(f'Subscribing to: {input_topic} -> publishing odom on: {output_topic}')
        # retrieved numeric value for noise standard deviation
        self.position_noise_std = float(self.get_parameter('position_noise_std').get_parameter_value().double_value)
        self.orientation_noise_std = float(self.get_parameter('orientation_noise_std').get_parameter_value().double_value)

    def pose_callback(self, pose_msg: Pose):
        # record offset on first message
        if not self.initialised:
            self.offset = deepcopy(pose_msg)
            self.initialised = True

        now_ns = self.get_clock().now().nanoseconds

        # on first message, initialize storage and zero velocity
        if self._prev_pose is None:
            self._prev_pose = deepcopy(pose_msg)
            self._prev_pose_time_ns = now_ns
            self.latest_vel = Twist()
            self.latest_pose = deepcopy(pose_msg)
            return

        dt = (now_ns - self._prev_pose_time_ns) / 1e9
        if dt > 0.0:
            vx = (pose_msg.position.x - self._prev_pose.position.x) / dt
            vy = (pose_msg.position.y - self._prev_pose.position.y) / dt
            vz = (pose_msg.position.z - self._prev_pose.position.z) / dt
            vel = Twist()
            vel.linear.x = vx
            vel.linear.y = vy
            vel.linear.z = vz
            vel.angular.x = 0.0
            vel.angular.y = 0.0
            vel.angular.z = 0.0
            self.latest_vel = vel

        # update stored pose and timestamp for next delta
        self._prev_pose = deepcopy(pose_msg)
        self._prev_pose_time_ns = now_ns
        self.latest_pose = deepcopy(pose_msg)

    def run(self):
        # publish odometry based on last stored pose and velocity at fixed rate
        if self.latest_pose is None:
            return

        odom = Odometry()
        odom.header.stamp = self.get_clock().now().to_msg()
        odom.header.frame_id = self.odom_frame
        odom.child_frame_id = self.base_frame

        # copy pose (don't mutate) and apply offset subtraction + gaussian noise
        odom.pose.pose = deepcopy(self.latest_pose)

        std = self.position_noise_std
        if std and std > 0.0:
            odom.pose.pose.position.x = (odom.pose.pose.position.x - self.offset.position.x) + random.gauss(0.0, std)
            odom.pose.pose.position.y = (odom.pose.pose.position.y - self.offset.position.y) + random.gauss(0.0, std)
            odom.pose.pose.position.z = 0.0

        std = self.orientation_noise_std
        if std and std > 0.0:
            q = odom.pose.pose.orientation
            q.x = (q.x - self.offset.orientation.x) + random.gauss(0.0, std)
            q.y = (q.y - self.offset.orientation.y) + random.gauss(0.0, std)
            q.z = (q.z - self.offset.orientation.z) + random.gauss(0.0, std)
            q.w = (q.w - self.offset.orientation.w) + random.gauss(0.0, std)
            mag = math.sqrt(q.x * q.x + q.y * q.y + q.z * q.z + q.w * q.w)
            if mag > 0.0:
                q.x /= mag
                q.y /= mag
                q.z /= mag
                q.w /= mag

        # default covariances (zeros); adjust if needed
        odom.pose.covariance = [0.0] * 36
        odom.twist.covariance = [0.0] * 36

        # attach last-computed velocity (or zero if not available)
        odom.twist.twist = deepcopy(self.latest_vel) if self.latest_vel is not None else Twist()

        self.odom_pub.publish(odom)


def main(args=None):
    rclpy.init(args=args)
    node = PoseToOdom()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/mobile_manip/src/t265_odom.cpp
#include <algorithm>
#include <cmath>
#include <cstddef>
#include <memory>
#include <random>
#include <string>

#include "geometry_msgs/msg/pose.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp/qos.hpp"

class PoseToOdom : public rclcpp::Node
{
public:
  PoseToOdom()
  : Node("t265_odom_publisher"),
    rng_(std::random_device{}())
  {
    this->declare_parameter<std::string>("input_topic", "sensors/gt");
    this->declare_parameter<std::string>("output_topic", "sensors/t265/pose/sample");
    this->declare_parameter<std::string>("odom_frame", "odom");
    this->declare_parameter<std::string>("base_frame", "t265_link");
    this->declare_parameter<double>("position_noise_std", 0.005);
    this->declare_parameter<double>("orientation_noise_std", 0.0001);
    this->declare_parameter<double>("publish_rate", 20.0);

    const auto input_topic = this->get_parameter("input_topic").as_string();
    const auto output_topic = this->get_parameter("output_topic").as_string();
    odom_frame_ = this->get_parameter("odom_frame").as_string();
    base_frame_ = this->get_parameter("base_frame").as_string();
    position_noise_std_ = this->get_parameter("position_noise_std").as_double();
    orientation_noise_std_ = this->get_parameter("orientation_noise_std").as_double();
    publish_rate_ = std::max(this->get_parameter("publish_rate").as_double(), 1e-6);

    const auto sensor_qos = rclcpp::SensorDataQoS().keep_last(1);

    odom_pub_ = this->create_publisher<nav_msgs::msg::Odometry>(output_topic, sensor_qos);
    pose_sub_ = this->create_subscription<geometry_msgs::msg::Pose>(
      input_topic,
      sensor_qos,
      std::bind(&PoseToOdom::pose_callback, this, std::placeholders::_1));
    timer_ = this->create_wall_timer(
      std::chrono::duration<double>(1.0 / publish_rate_),
      std::bind(&PoseToOdom::run, this));

    RCLCPP_INFO(
      this->get_logger(),
      "Subscribing to: %s -> publishing odom on: %s",
      input_topic.c_str(),
      output_topic.c_str());
  }

private:
  struct RollPitchYaw
  {
    double roll;
    double pitch;
    double yaw;
  };

  static double normalize_angle(double angle)
  {
    return std::atan2(std::sin(angle), std::cos(angle));
  }

  static double heading_from_quaternion(const geometry_msgs::msg::Quaternion & q)
  {
    const double siny_cosp = 2.0 * (q.w * q.z + q.x * q.y);
    const double cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z);
    return std::atan2(siny_cosp, cosy_cosp);
  }

  static RollPitchYaw rpy_from_quaternion(const geometry_msgs::msg::Quaternion & q)
  {
    constexpr double kHalfPi = 1.5707963267948966;
    RollPitchYaw rpy{};

    const double sinr_cosp = 2.0 * (q.w * q.x + q.y * q.z);
    const double cosr_cosp = 1.0 - 2.0 * (q.x * q.x + q.y * q.y);
    rpy.roll = std::atan2(sinr_cosp, cosr_cosp);

    const double sinp = 2.0 * (q.w * q.y - q.z * q.x);
    if (std::abs(sinp) >= 1.0) {
      rpy.pitch = std::copysign(kHalfPi, sinp);
    } else {
      rpy.pitch = std::asin(sinp);
    }

    rpy.yaw = heading_from_quaternion(q);
    return rpy;
  }

  static geometry_msgs::msg::Quaternion quaternion_from_rpy(
    double roll,
    double pitch,
    double yaw)
  {
    geometry_msgs::msg::Quaternion q;

    const double cy = std::cos(yaw * 0.5);
    const double sy = std::sin(yaw * 0.5);
    const double cp = std::cos(pitch * 0.5);
    const double sp = std::sin(pitch * 0.5);
    const double cr = std::cos(roll * 0.5);
    const double sr = std::sin(roll * 0.5);

    q.w = cr * cp * cy + sr * sp * sy;
    q.x = sr * cp * cy - cr * sp * sy;
    q.y = cr * sp * cy + sr * cp * sy;
    q.z = cr * cp * sy - sr * sp * cy;
    return q;
  }

  double gaussian(double stddev)
  {
    if (stddev <= 0.0) {
      return 0.0;
    }
    std::normal_distribution<double> dist(0.0, stddev);
    return dist(rng_);
  }

  void pose_callback(const geometry_msgs::msg::Pose::SharedPtr msg)
  {
    const auto now = this->get_clock()->now();

    if (!initialized_) {
      offset_ = *msg;
      initialized_ = true;
    }

    if (!have_prev_pose_) {
      prev_pose_ = *msg;
      prev_yaw_ = heading_from_quaternion(msg->orientation);
      prev_pose_time_ = now;
      latest_vel_ = geometry_msgs::msg::Twist();
      latest_pose_ = *msg;
      have_prev_pose_ = true;
      have_latest_pose_ = true;
      return;
    }

    const double dt = (now - prev_pose_time_).seconds();
    if (dt > 0.0) {
      const double yaw = heading_from_quaternion(msg->orientation);
      latest_vel_.linear.x = (msg->position.x - prev_pose_.position.x) / dt;
      latest_vel_.linear.y = (msg->position.y - prev_pose_.position.y) / dt;
      latest_vel_.linear.z = (msg->position.z - prev_pose_.position.z) / dt;
      latest_vel_.angular.x = 0.0;
      latest_vel_.angular.y = 0.0;
      latest_vel_.angular.z = normalize_angle(yaw - prev_yaw_) / dt;
      prev_yaw_ = yaw;
    }

    prev_pose_ = *msg;
    prev_pose_time_ = now;
    latest_pose_ = *msg;
    have_latest_pose_ = true;
  }

  void run()
  {
    if (!have_latest_pose_) {
      return;
    }

    nav_msgs::msg::Odometry odom;
    odom.header.stamp = this->get_clock()->now();
    odom.header.frame_id = odom_frame_;
    odom.child_frame_id = base_frame_;
    odom.pose.pose = latest_pose_;

    if (position_noise_std_ > 0.0) {
      odom.pose.pose.position.x =
        (odom.pose.pose.position.x - offset_.position.x) + gaussian(position_noise_std_);
      odom.pose.pose.position.y =
        (odom.pose.pose.position.y - offset_.position.y) + gaussian(position_noise_std_);
      odom.pose.pose.position.z = 0.0;
    }

    {
      const auto rpy = rpy_from_quaternion(odom.pose.pose.orientation);
      const double yaw_offset = heading_from_quaternion(offset_.orientation);
      const double yaw_noise =
        orientation_noise_std_ > 0.0 ? gaussian(orientation_noise_std_) : 0.0;
      odom.pose.pose.orientation = quaternion_from_rpy(
        rpy.roll,
        rpy.pitch,
        normalize_angle(rpy.yaw - yaw_offset + yaw_noise));
    }

    odom.twist.twist = latest_vel_;
    odom_pub_->publish(odom);
  }

  std::string odom_frame_;
  std::string base_frame_;
  double position_noise_std_{0.0};
  double orientation_noise_std_{0.0};
  double publish_rate_{20.0};

  bool initialized_{false};
  bool have_prev_pose_{false};
  bool have_latest_pose_{false};
  geometry_msgs::msg::Pose offset_;
  geometry_msgs::msg::Pose latest_pose_;
  geometry_msgs::msg::Pose prev_pose_;
  geometry_msgs::msg::Twist latest_vel_;
  double prev_yaw_{0.0};
  rclcpp::Time prev_pose_time_{0, 0, RCL_ROS_TIME};
  std::mt19937 rng_;

  rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_pub_;
  rclcpp::Subscription<geometry_msgs::msg::Pose>::SharedPtr pose_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<PoseToOdom>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
